### PR TITLE
feat: [PL-28164]: Listing Pages: Preserve Search Context & Open in new tab

### PR DESCRIPTION
**Pull request by mayankVermaHarness**
### Summary

JIRA: https://harness.atlassian.net/browse/PL-28164
Listing Pages: Preserving Search Context & Open in new tab

#### Screenshots

Open in new tab (same for Users/UserGroups/ServiceAccounts/ResourceGroups/Roles):
image

Preserve Search Context in query Params (same for Users/UserGroups/ServiceAccounts/ResourceGroups/Roles):
image

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

#### [Contributor license agreement](https://github.com/harness/harness-core-ui/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

[PL-28164]: https://harness.atlassian.net/browse/PL-28164?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by mayankVermaHarness** [comment]
retrigger cypress-pipeline

File: src/modules/10-common/hooks/useStateWithQueryParams.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+  useEffect(() => {
+    // eslint-disable-next-line
+    // @ts-ignore
+    const valueFromQP = qp[key]
+    if (valueFromQP) {
+      setStateVal?.(valueFromQP)
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
```
Comment (abhinavrastogi-harness) [inline src/modules/10-common/hooks/useStateWithQueryParams.ts:33]: we should allow override the dependencies array

File: src/modules/10-common/hooks/useStateWithQueryParams.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+}
+export function useStateWithQueryParams(options: UseSyncQueryParamsOptions): [string, (v: string) => void] {
+  const { key, defaultVal = '' } = options
+  const [stateVal, setStateVal] = useState<string>(defaultVal)
+  const { updateQueryParams } = useUpdateQueryParams()
+  const qp = useQueryParams()
+
+  useEffect(() => {
+    // eslint-disable-next-line
+    // @ts-ignore
```
Comment (abhinavrastogi-harness) [inline src/modules/10-common/hooks/useStateWithQueryParams.ts:27]: why do we need ts-ignore?

File: src/framework/utils/WindowLocation.ts [hunk 3]
```diff
@@ -1,8 +1,26 @@
...
+export const getLocationPathName = (): string => {
   let pathName = window.location.pathname
   if (window.browserRouterEnabled) {
     const nameSpace = window.harnessNameSpace ? `${window.harnessNameSpace}/` : ''
     pathName = `/${nameSpace}ng/`
   }
   return pathName
 }
+
+export const getOriginWithPath = (): string => {
```
Comment (avinashmadhwani02) [inline src/framework/utils/WindowLocation.ts:17]: Add a test case for this util method.

File: src/modules/10-common/hooks/useStateWithQueryParams.ts (hunk 2 above)
Comment (avinashmadhwani02) [reply src/modules/10-common/hooks/useStateWithQueryParams.ts:27]: +1

File: src/modules/10-common/hooks/useStateWithQueryParams.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    const valueFromQP = qp[key]
+    if (valueFromQP) {
+      setStateVal?.(valueFromQP)
```
Comment (avinashmadhwani02) [inline src/modules/10-common/hooks/useStateWithQueryParams.ts:30]: Do we need `?` when setting state?

File: src/modules/10-common/hooks/useStateWithQueryParams.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * that can be found in the licenses directory at the root of this repository, also available at
+ * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
+ */
+
+import { useEffect, useState } from 'react'
+import type { IParseOptions } from 'qs'
+import { useQueryParams, useUpdateQueryParams } from '@common/hooks'
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+// type NotFunction<T> = T extends Function ? never : T
```
Comment (avinashmadhwani02) [inline src/modules/10-common/hooks/useStateWithQueryParams.ts:13]: Remove comments?

File: src/modules/20-rbac/components/ResourceGroupList/ResourceGroupColumnMenu.tsx [hunk 6]
```diff
@@ -108,6 +119,17 @@ const ResourceGroupColumnMenu: Renderer<CellPropsResourceGroupColumn<ResourceGro
...
       />
       <Menu style={{ minWidth: 'unset' }}>
+        <li>
+          <Link
+            className={cx('bp3-menu-item', css.openNewTabStyle)}
+            target="_blank"
+            to={resourceGroupDetailsUrl}
+            onClick={e => e.stopPropagation()}
+          >
+            <Icon name="launch" style={{ marginRight: 'var(--spacing-small)' }} />
```
Comment (avinashmadhwani02) [inline src/modules/20-rbac/components/ResourceGroupList/ResourceGroupColumnMenu.tsx:129]: avoid inline styling

File: src/modules/20-rbac/components/RoleCard/RoleCard.tsx [hunk 7]
```diff
@@ -85,20 +86,37 @@ const RoleCard: React.FC<RoleCardProps> = ({ data, reloadRoles, editRoleModal })
...
         menuContent={
           <Menu>
+            <li>
+              <Link
+                className={cx('bp3-menu-item', css.openNewTabStyle)}
+                target="_blank"
+                to={roleDetailsUrl}
+                onClick={e => e.stopPropagation()}
+              >
+                <Icon name="launch" style={{ marginRight: 'var(--spacing-small)' }} />
```
Comment (avinashmadhwani02) [inline src/modules/20-rbac/components/RoleCard/RoleCard.tsx:116]: avoid inline styling

File: src/modules/20-rbac/pages/ResourceGroups/views/ResourceGroupsList.tsx [hunk 8]
```diff
@@ -14,11 +14,13 @@ import { useGetResourceGroupListV2 } from 'services/resourcegroups'
 import ResourceGroupListView from '@rbac/components/ResourceGroupList/ResourceGroupListView'
 import type { PipelineType, ProjectPathProps } from '@common/interfaces/RouteInterfaces'
 import { useDocumentTitle } from '@common/hooks/useDocumentTitle'
+import { useStateWithQueryParams } from '@common/hooks/useStateWithQueryParams'
 import RbacButton from '@rbac/components/Button/Button'
 import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
 import { ResourceType } from '@rbac/interfaces/ResourceType'
 import useRBACError from '@rbac/utils/useRBACError/useRBACError'
 import routes from '@common/RouteDefinitions'
+// import { useQueryParams, useUpdateQueryParams } from '@common/hooks'
```
Comment (avinashmadhwani02) [inline src/modules/20-rbac/pages/ResourceGroups/views/ResourceGroupsList.tsx:23]: Remove comments

File: src/modules/20-rbac/pages/ResourceGroups/views/ResourceGroupsList.tsx [hunk 9]
```diff
@@ -95,6 +97,8 @@ const ResourceGroupsList: React.FC = () => {
         toolbar={
           <Layout.Horizontal margin={{ right: 'small' }} height="xxxlarge">
             <ExpandingSearchInput
+              // saveInQueryParams
```
Comment (avinashmadhwani02) [inline src/modules/20-rbac/pages/ResourceGroups/views/ResourceGroupsList.tsx:100]: comments

File: src/modules/10-common/hooks/useStateWithQueryParams.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+export interface UseSyncQueryParamsOptions extends IParseOptions {
+  key: string
+  defaultVal?: string
+}
+export function useStateWithQueryParams(options: UseSyncQueryParamsOptions): [string, (v: string) => void] {
+  const { key, defaultVal = '' } = options
```
Comment (avinashmadhwani02) [inline src/modules/10-common/hooks/useStateWithQueryParams.ts:20]: What is the use of defaultValue?
If we are assigning it to state, it should be updated in the query param also. 
both should be in sync

File: src/modules/20-rbac/pages/ServiceAccounts/views/ServiceAccountsListView.tsx [hunk 11]
```diff
@@ -203,6 +214,17 @@ const RenderColumnMenu: Renderer<CellProps<ServiceAccountAggregateDTO>> = ({ row
           }}
         />
         <Menu>
+          <li>
+            <Link
```
Comment (avinashmadhwani02) [inline src/modules/20-rbac/pages/ServiceAccounts/views/ServiceAccountsListView.tsx:218]: This code is repeated 3-4 times. Can we try to re-use it?

File: src/modules/10-common/hooks/useStateWithQueryParams.ts (hunk 4 above)
Comment (mayankVermaHarness) [reply src/modules/10-common/hooks/useStateWithQueryParams.ts:30]: No. Fixed it

File: src/modules/10-common/hooks/useStateWithQueryParams.ts (hunk 1 above)
Comment (mayankVermaHarness) [reply src/modules/10-common/hooks/useStateWithQueryParams.ts:33]: As discussed, changed useEffect to allow for this.

Although, the initial reason this hook was written was for `<ExpandingSearchInput />`—but this component does not update its internal state on the basis of changed `defaultValue` from outside. It only sets `defaultValue` once on mount—this means even though state will update in parent component, it won't update internally in `<ExpandingSearchInput />`.

This is a problem of `<ExpandingSearchInput />`, but now `useStateWithQueryParams` will work the way it's supposed to.

Will discuss this further on a call.

File: src/modules/20-rbac/components/ResourceGroupList/ResourceGroupColumnMenu.tsx (hunk 6 above)
Comment (mayankVermaHarness) [reply src/modules/20-rbac/components/ResourceGroupList/ResourceGroupColumnMenu.tsx]: done

File: src/modules/20-rbac/pages/ResourceGroups/views/ResourceGroupsList.tsx (hunk 8 above)
Comment (mayankVermaHarness) [reply src/modules/20-rbac/pages/ResourceGroups/views/ResourceGroupsList.tsx]: done

File: src/modules/20-rbac/pages/ResourceGroups/views/ResourceGroupsList.tsx (hunk 9 above)
Comment (mayankVermaHarness) [reply src/modules/20-rbac/pages/ResourceGroups/views/ResourceGroupsList.tsx]: done

File: src/modules/20-rbac/pages/ServiceAccounts/views/ServiceAccountsListView.tsx (hunk 11 above)
Comment (mayankVermaHarness) [reply src/modules/20-rbac/pages/ServiceAccounts/views/ServiceAccountsListView.tsx]: Done, created a new component `OpenInNewTab`

File: src/modules/10-common/hooks/useStateWithQueryParams.ts (hunk 10 above)
Comment (mayankVermaHarness) [reply src/modules/10-common/hooks/useStateWithQueryParams.ts:18]: Done

File: src/modules/10-common/hooks/useStateWithQueryParams.ts (hunk 2 above)
Comment (mayankVermaHarness) [reply src/modules/10-common/hooks/useStateWithQueryParams.ts]: Removed, had to add type to `useQueryParam` for this.

**Comment by mayankVermaHarness** [comment]
retrigger jest

**Comment by mayankVermaHarness** [comment]
retrigger coverage

**Comment by mayankVermaHarness** [comment]
retrigger coverage

**Review by abhinavrastogi-harness — APPROVED** [review]

**Review by vkbansal-harness — APPROVED** [review]

**Pull request merged**